pkg/auth/providerrefresh/refresher.go

```diff
@@ -277,6 +277,20 @@ func (r *refresher) refreshAttributes(attribs *v3.UserAttribute) (*v3.UserAttrib
 			}
 			userPrincipal, err := providers.GetPrincipal(principalID, token)
 			if err != nil {
+				// If the token is no longer valid, delete the token
```

Comment:
I agree that this change could be erroneous because we are relying on error messages that can be changed in future. However I couldn't find a better way to identify inactive tokens and there are quite a few places in Rancher code that looks for certain error messages.

To answer your questions:

> Check why providers.CanAccessWithGroupProviders at https://github.com/verrazzano/rancher/blob/51373799ed86269cdae03aa3f9e6f1c2ef1c177d/pkg/auth/providerrefresh/refresher.go#L257C56-L257C56 is returning true or not returning an error, should it either return false or error so that the flow does not reach here ?

The `providers.CanAccessWithGroupProviders`  function mainly checks that the user still has Rancher access. It validates the user permissions configured in the Rancher's auth provider config. This does not validate anything from the Keycloak side. This is of not use for the use case that we are trying to solve.

> Can we look at instead setting a reasonable value for [auth-user-session-ttl-minutes](https://github.com/verrazzano/rancher/blob/oracle/release/2.7.3/pkg/settings/setting.go#L143) by updating the setting while install/upgrade so that the [purge_daemon](https://github.com/verrazzano/rancher/blob/51373799ed86269cdae03aa3f9e6f1c2ef1c177d/pkg/auth/tokens/purge_daemon.go#L44C1-L44C1) takes care of cleanup rather than we doing it based on some error message

Yes, I started with  [auth-user-session-ttl-minutes](https://github.com/verrazzano/rancher/blob/oracle/release/2.7.3/pkg/settings/setting.go#L143) settings itself but due to the following two issues, I think this may not be the right solution.
- Keycloak SSO idle timeout is set to 30 minutes by default. If we set  [auth-user-session-ttl-minutes](https://github.com/verrazzano/rancher/blob/oracle/release/2.7.3/pkg/settings/setting.go#L143) also to 30 minutes or some reasonable value, then the user gets logged out even when he is actively accessing the Rancher console.
- For some reason, if the auth session is deleted/invalidated at the Keycloak side before [auth-user-session-ttl-minutes](https://github.com/verrazzano/rancher/blob/oracle/release/2.7.3/pkg/settings/setting.go#L143), we still see the same issue.